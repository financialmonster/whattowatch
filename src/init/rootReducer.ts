import { connectRouter } from 'connected-react-router/immutable';
import { combineReducers } from 'redux';
import { createBrowserHistory } from 'history';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import immutableTransform from 'redux-persist-transform-immutable';

import authReducer from 'domains/auth/authReducer';
import promoReducer from 'domains/promo/promoReducer';
import filmsReducer from 'domains/films/filmsReducer';
import reviewsReducer from 'domains/reviews/reviewsReducer';
import favoritesReducer from 'domains/favorites/favoritesReducer';

const persistConfig = {
    transforms: [immutableTransform()],
    key: `root`,
    storage
}

export const history = createBrowserHistory();
const routerReducer = connectRouter(history);

const rootReducer = combineReducers({
    router: routerReducer,
    auth: authReducer,
    promo: promoReducer,
    films: filmsReducer,
    reviews: reviewsReducer,
    favorites: favoritesReducer
});

export const persistedRootReducer = persistReducer(persistConfig, rootReducer);
