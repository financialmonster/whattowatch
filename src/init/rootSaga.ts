import { all, call } from 'redux-saga/effects';
import { SagaIterator } from '@redux-saga/core';

import { watchPromo } from 'domains/promo/saga/promoWatchers';
import { watchFilms } from 'domains/films/saga/filmsWatchers';
import { watchAuth } from 'domains/auth/saga/authWatchers';
import { watchReviews } from 'domains/reviews/saga/reviewsWatchers';

export function* rootSaga(): SagaIterator {
    yield all([call( watchPromo ), call( watchFilms ), call( watchAuth ), call( watchReviews )]);
}
