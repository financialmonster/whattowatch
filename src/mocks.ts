import { Map, fromJS, List } from 'immutable';

export const createStoreMock = (mode: string = `common`) => {
    return {
        promo: Map({
            promo: Map(),
            isPromoFetching: true,
            promoError: null
        }),
        films: Map({
            films: fromJS([{genre: `Thriller`, id: 1}, {genre: `Comedy`, id: 2}]),
            isFilmsFetching: mode === `detailed`,
            filmsError: null,
            filter: `All genres`
        }),
        auth: Map({
            user: null,
            authError: null,
            isAuthFetching: false,
            authStatusError: null,
            isAuthStatusFetching: false
        }),
        reviews: Map({
            reviews: List(),
            isReviewsFetching: true,
            reviewsError: null
        })
    }
}
