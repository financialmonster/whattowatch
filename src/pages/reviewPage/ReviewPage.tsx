import React, { FC } from 'react';
import { Redirect, useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { Map } from 'immutable';

import * as authSelectors from 'domains/auth/authSelectors';
import { Routes } from 'mainConstants';
import { Header } from 'components/header/Header';
import { Footer } from 'components/footer/Footer';
import * as filmsSelectors from 'domains/films/filmsSelectors';
import { isFilms } from 'utils';
import { ReviewFormContainer } from 'components/reviewFormContainer/ReviewFormContainer';

export const ReviewPage: FC = () => {
    const user = useSelector(authSelectors.getUser);
    const films = useSelector(filmsSelectors.getFilms);
    const {id} = useParams();

    let film;
    let sectionStyle = {};    

    if(isFilms(films)) {
        film = films.find((film: Map<String, any>) => film.get(`id`) === +id);

        const backgroundColor: string = (film as Map<string, any>).get(`background_color`);

        sectionStyle = { backgroundColor };
    }

    if(!user) {
        return <Redirect to={Routes.LOGIN_PAGE} />;
    }

    return (
        <section className="movie-card movie-card--full" style={ sectionStyle }>
            <div className="movie-card__header">
                <Header film={ film } />
            </div>
            <div className="add-review">
                <ReviewFormContainer />
            </div>
            <Footer />
        </section>
    );
}
