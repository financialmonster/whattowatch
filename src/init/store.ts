import { createStore, applyMiddleware } from 'redux';
import { createLogger } from 'redux-logger';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { routerMiddleware as createRouterMiddleware } from 'connected-react-router';

import { rootReducer, history } from './rootReducer';

const routerMiddleware = createRouterMiddleware(history);
const middlewares = [ routerMiddleware ];

const logger = createLogger({
    duration: true,
    collapsed: true,
    colors: {
        title: (action) => (action.error) ? 'firebrick' : 'deepskyblue',
        prevState: () => '#1C5FAF',
        action:    () => '#149945',
        nextState: () => '#A47104',
        error:     () => '#ff0005',
    }
});

if(process.env.NODE_ENV !== 'production') {
    middlewares.push(logger);
}

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(...middlewares)));

export default store;
