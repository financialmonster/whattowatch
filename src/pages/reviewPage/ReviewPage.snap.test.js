/* eslint-disable import/first */
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({id: 1})
}));

import React from 'react';
import renderer from 'react-test-renderer';
import { Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { routerMiddleware as createRouterMiddleware } from 'connected-react-router';

import { ReviewPage } from './ReviewPage';
import { createStoreMock } from 'mocks';
import { history } from 'init/rootReducer'; 

it(`ReviewPage: correctly rendered`, () => {
    const routerMiddleware = createRouterMiddleware(history);
    const mockStore = configureStore([ routerMiddleware ]);
    const store = mockStore(createStoreMock());

    const tree = renderer.create(
        <Provider store={store}>
            <Router history={history}>
                <ReviewPage />
            </Router>
        </Provider>        
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
