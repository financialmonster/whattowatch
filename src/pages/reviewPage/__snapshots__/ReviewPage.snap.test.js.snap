// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewPage: correctly rendered 1`] = `null`;
